File: backend/agent/schemas.py
from pydantic import BaseModel, Field
from typing import List, Optional, Literal

class FrictionMoment(BaseModel):
    id: str
    timestamp: str
    categories: List[str]
    textNote: Optional[str] = None
    voiceNoteBase64: Optional[str] = None
    voiceNoteDuration: Optional[float] = None
    ambientLevel: Optional[float] = None
    classContext: Optional[str] = None

class SupportPreference(BaseModel):
    id: str
    label: str
    enabled: bool
    custom: bool

class StructuredFrictionEvent(BaseModel):
    environmentalFactors: List[str] = Field(description="Specific environmental triggers, e.g. 'loud cafeteria', 'substitute teacher'")
    categories: List[str] = Field(description="Best matching categories")
    contextSignals: List[str] = Field(description="Important context, e.g. 'Monday morning', 'Math class'")
    uncertaintySignals: List[str] = Field(description="Things that were unclear or unexpected")
    socialContext: Optional[str] = Field(default=None, description="Social environment details")
    supportMentioned: Optional[List[str]] = Field(default=None, description="Any supports the student mentioned they used or needed")
    confidence: float = Field(description="Confidence in this extraction", ge=0, le=1)

class EvidenceComparison(BaseModel):
    conditionA: str
    conditionB: str
    observedDifference: Optional[str] = None

class EvidenceBackedPattern(BaseModel):
    statement: str = Field(description="The core pattern statement, e.g. '3 of 4 high-friction moments followed unexpected schedule changes.'")
    supportingMomentIds: List[str] = Field(description="IDs of moments supporting this pattern")
    contradictoryMomentIds: Optional[List[str]] = Field(default=None, description="IDs of moments contradicting this pattern")
    sampleSize: int
    evidenceStrength: Literal["insufficient", "weak", "moderate", "strong"]
    comparison: Optional[EvidenceComparison] = Field(default=None, description="Optional comparison if evaluating an accommodation or specific context")
    limitations: List[str] = Field(description="Limitations of this pattern, e.g. 'Only 3 data points'")
    languageMode: Literal["observation", "association"] = Field(description="Must be observation or association, NEVER causal or diagnostic")

class EvidenceAssessment(BaseModel):
    strength: Literal["insufficient", "weak", "moderate", "strong"]
    supportingCount: int
    contradictoryCount: int
    confidence: float = Field(ge=0, le=1)
    limitations: List[str]
    recommendation: Literal["do_not_surface", "surface_as_observation"]

class FrictionInsight(BaseModel):
    summary: str
    patterns: List[EvidenceBackedPattern]
    helpfulPreferences: List[str]

class SafetyAssessment(BaseModel):
    isSafe: bool
    flags: List[str] = Field(description="Any safety violations found")
    reasoning: str

File: src/agent/schemas.ts
import { z } from 'zod';

export const structuredFrictionEventSchema = z.object({
  environmentalFactors: z.array(z.string()).describe("Specific environmental triggers, e.g. 'loud cafeteria', 'substitute teacher'"),
  categories: z.array(z.string()).describe("Best matching categories"),
  contextSignals: z.array(z.string()).describe("Important context, e.g. 'Monday morning', 'Math class'"),
  uncertaintySignals: z.array(z.string()).describe("Things that were unclear or unexpected"),
  socialContext: z.string().optional().describe("Social environment details"),
  supportMentioned: z.array(z.string()).optional().describe("Any supports the student mentioned they used or needed"),
  confidence: z.number().min(0).max(1).describe("Confidence in this extraction"),
});

export const evidenceBackedPatternSchema = z.object({
  statement: z.string().describe("The core pattern statement, e.g. '3 of 4 high-friction moments followed unexpected schedule changes.'"),
  supportingMomentIds: z.array(z.string()).describe("IDs of moments supporting this pattern"),
  contradictoryMomentIds: z.array(z.string()).optional().describe("IDs of moments contradicting this pattern"),
  sampleSize: z.number(),
  evidenceStrength: z.enum(["insufficient", "weak", "moderate", "strong"]),
  comparison: z.object({
    conditionA: z.string(),
    conditionB: z.string(),
    observedDifference: z.string().optional(),
  }).optional(),
  limitations: z.array(z.string()).describe("Limitations of this pattern, e.g. 'Only 3 data points'"),
  languageMode: z.enum(["observation", "association"]).describe("Must be observation or association, NEVER causal or diagnostic"),
});

export const evidenceAssessmentSchema = z.object({
  strength: z.enum(["insufficient", "weak", "moderate", "strong"]),
  supportingCount: z.number(),
  contradictoryCount: z.number(),
  confidence: z.number().min(0).max(1),
  limitations: z.array(z.string()),
  recommendation: z.enum(["do_not_surface", "surface_as_observation"]),
});

export const frictionInsightSchema = z.object({
  summary: z.string(),
  patterns: z.array(evidenceBackedPatternSchema),
  helpfulPreferences: z.array(z.string()),
});

export const safetyAssessmentSchema = z.object({
  isSafe: z.boolean(),
  flags: z.array(z.string()).describe("Any safety violations found"),
  reasoning: z.string(),
});
